include/aws/crt/Variant.h

```diff
@@ -0,0 +1,595 @@
+                    std::tuple<typename std::add_pointer<Ts>::type...> as_tuple;
+
+                  private:
+                    template <short Index, typename First, typename Second, typename... Rest>
+                    void InitTuple(char *storage)
+                    {
+                        First *value = reinterpret_cast<First *>(storage);
+                        std::get<Index>(as_tuple) = value;
+                        InitTuple<Index + 1, Second, Rest...>(storage);
+                    }
+
+                    template <short Index, typename Last> void InitTuple(char *storage)
+                    {
+                        Last *value = reinterpret_cast<Last *>(storage);
+                        std::get<Index>(as_tuple) = value;
+                    }
+                };
+            } // namespace VariantDebug
+#endif        /* defined(DEBUG_BUILD) */
+        }     // namespace VariantDetail
+
+        template <std::size_t Index, typename... Ts> class VariantAlternative;
+
+        template <typename T> struct VariantInPlaceInitT
+        {
+            explicit VariantInPlaceInitT() = default;
+        };
+
+        /**
+         * Custom implementation of a Variant type. std::variant requires C++17
+         * @tparam Ts types of the variant value
+         */
+        template <typename... Ts> class Variant
+        {
+          public:
+            static constexpr std::size_t AlternativeCount = sizeof...(Ts);
+
+            template <std::size_t Index> using ThisVariantAlternative = VariantAlternative<Index, Ts...>;
+
+            using ThisVariantT = Variant<Ts...>;
+            template <typename OtherT>
+            using EnableIfOtherIsSameVariantType =
+                typename std::enable_if<std::is_same<typename std::decay<OtherT>, ThisVariantT>::value, int>::type;
+
+            template <typename OtherT>
+            using EnableIfOtherIsThisVariantAlternative = typename std::
+                enable_if<VariantDetail::Checker::HasType<typename std::decay<OtherT>::type, Ts...>::value, int>::type;
+
+            Variant()
+            {
+                using FirstAlternative = typename ThisVariantAlternative<0>::type;
+                new (m_storage) FirstAlternative();
+                m_index = 0;
+            }
+
+            Variant(const Variant &other)
+            {
+                auto otherIdx = other.m_index;
+                if (otherIdx != -1)
+                {
+                    CopyConstructUtil<0, Ts...>(other);
+                }
+            }
+
+            Variant(Variant &&other)
+            {
+                auto otherIdx = other.m_index;
+                if (otherIdx != -1)
+                {
+                    MoveConstructUtil<0, Ts...>(std::move(other));
+                }
+            }
+
+            template <typename T, EnableIfOtherIsThisVariantAlternative<T> = 1> Variant(const T &val)
```

Comment:
Debatable: This constructor is not needed, as the constructor with a forwarding reference parameter covers const references.

Reply:
Good catch, thank you!
However, my brain hurts more trying to think more about this one. I'd prefer to keep it as is as of now.